### PR TITLE
fix(stagewise): implement gitignore-aware filtering and workspace root resolution for agent edits

### PR DESCRIPTION
**Pull request by juliangoetze**
## Summary by CodeRabbit

* **New Features**
  * Workspace-aware .gitignore filtering for edit tracking, optional per-edit workspace context, and per-tool-call fan-out caps that emit a single coarse telemetry event when hit.
  * Toolbox: access to mounted-paths and improved owning-workspace selection for nested mounts and cross-workspace operations.

* **Tests**
  * Expanded tests covering gitignore semantics (negations, nesting, fallbacks), fan-out cap/telemetry behavior, and workspace-mount resolution edge cases.

### CODE REVIEW

File: apps/browser/src/backend/services/diff-history/index.ts [hunk 1]
```diff
@@ -323,6 +390,118 @@ export class DiffHistoryService extends DisposableService {
...
+    filepath: string,
+    workspaceRoot?: string | null,
+  ): Promise<boolean> {
+    try {
+      // Phase 1: synchronous segment denylist. Hot path for runaway
+      // node_modules fan-out; rejects in microseconds.
+      const segments = filepath.split(path.sep);
+      for (const seg of segments) {
+        if (HARDCODED_DENY_SEGMENTS.has(seg)) return false;
+      }
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/diff-history/index.ts:552]: Good one, adding a fix.

File: apps/browser/src/backend/services/diff-history/index.ts [hunk 2]
```diff
@@ -338,10 +517,56 @@ export class DiffHistoryService extends DisposableService {
...
+          `[DiffHistory] Tool call ${edit.toolCallId} exceeded fan-out cap ` +
+            `(${this.MAX_EDITS_PER_TOOL_CALL} edits). Further edits in ` +
+            `this call are dropped from diff history; on-disk writes are unaffected.`,
+        );
+        this.telemetryService.capture('diff-history-fanout-cap-hit', {
+          tool_call_id: edit.toolCallId,
+          agent_instance_id: edit.agentInstanceId,
+          first_dropped_path: edit.path,
+          cap: this.MAX_EDITS_PER_TOOL_CALL,
+        });
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/diff-history/index.ts:640]: Good catch, will push a fix

File: apps/browser/src/backend/services/toolbox/index.ts [hunk 3]
```diff
@@ -444,6 +454,11 @@ export class ToolboxService extends DisposableService {
             agentInstanceId,
             path: absolutePath,
             toolCallId,
+            workspaceRoot:
+              this.mountManagerService?.findWorkspaceForFile(
+                agentInstanceId,
+                absolutePath,
+              ) ?? null,
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/toolbox/index.ts:461]: Good catch, adding a fix

File: apps/browser/src/backend/services/diff-history/index.ts [hunk 4]
```diff
@@ -247,6 +365,12 @@ export class DiffHistoryService extends DisposableService {
       })
       .on('change', async (path) => {
         if (this.filesIgnoredByWatcher.has(path)) return;
+        // Defense-in-depth: ignore changes in gitignored paths (e.g.
+        // node_modules, build dirs). The watcher should only be
+        // tracking files we deliberately added via `ensureWatched`,
+        // but a bad op from a prior session could leave an ignored
+        // file in the watch set — filter anyway.
+        if (!(await this.shouldTrackFilepath(path))) return;
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/diff-history/index.ts]: Pushing a fix

File: apps/browser/src/backend/services/diff-history/index.ts [hunk 5]
```diff
@@ -323,6 +448,118 @@ export class DiffHistoryService extends DisposableService {
...
+  private findWorkspaceRoot(filepath: string): string | null {
+    const mounts = this._mountPathsResolver?.();
+    if (!mounts || mounts.size === 0) return null;
+    let best: string | null = null;
+    for (const root of mounts) {
+      if (filepath === root || filepath.startsWith(root + path.sep)) {
+        if (!best || root.length > best.length) best = root;
+      }
+    }
+    return best;
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/diff-history/index.ts]: Good catch ,pusing a fix

File: apps/browser/src/backend/services/toolbox/services/mount-manager/index.ts [hunk 6]
```diff
@@ -39,6 +39,44 @@ function mountPrefixForPath(workspacePath: string): MountPrefix {
...
+  candidates: Iterable<string>,
+): string | undefined {
+  let best: string | undefined;
+  for (const wsPath of candidates) {
+    const isExactMatch = filePath === wsPath;
+    const isDescendant = filePath.startsWith(wsPath + path.sep);
+    if (!isExactMatch && !isDescendant) continue;
+    if (!best || wsPath.length > best.length) best = wsPath;
+  }
+  return best;
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/toolbox/services/mount-manager/index.ts]: Already fixed

File: apps/browser/src/backend/services/diff-history/index.test.ts [hunk 7]
```diff
@@ -2993,4 +2993,481 @@ describe('DiffHistoryService (E2E)', () => {
...
+        CAP,
+      );
+      expect(warnSpy).toHaveBeenCalledTimes(1);
+      expect(
+        (mockTelemetry.capture as ReturnType<typeof vi.fn>).mock.calls.filter(
+          ([event]) => event === 'diff-history-fanout-cap-hit',
+        ),
+      ).toHaveLength(1);
+
+      innerLogger.warn = originalWarn;
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/diff-history/index.test.ts]: Added a fix

File: apps/browser/src/backend/utils/load-gitignore.ts [hunk 8]
```diff
@@ -3,27 +3,107 @@ import { readFile } from 'node:fs/promises';
...
+    // Log files.
+    '**/*.log',
+  ]);
+
+  // Project's `.gitignore` added LAST — its rules (including
+  // negations like `!dist/keep.ts`) win over the soft defaults.
+  const gitignorePath = join(dir, '.gitignore');
   if (existsSync(gitignorePath)) {
     const content = await readFile(gitignorePath, 'utf-8');
     ig.add(content);
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/utils/load-gitignore.ts]: Adding a fix!

File: apps/browser/src/backend/services/toolbox/index.ts [hunk 9]
```diff
@@ -303,6 +303,16 @@ export class ToolboxService extends DisposableService {
...
 
+  /**
+   * Narrow accessor used by `DiffHistoryService` to resolve which
+   * filepaths belong to a mounted workspace (for the gitignore check).
+   * Returns an empty set before the mount manager has finished its
+   * async initialization — callers must tolerate that window.
+   */
+  public getAllMountedPaths(): Set<string> {
+    return this.mountManagerService?.getAllMountedPaths() ?? new Set();
+  }
```
[the remark this answers is not in the document]
Comment (juliangoetze) [reply apps/browser/src/backend/services/toolbox/index.ts:314]: Pushing a fix
